segmentstore/server/src/main/java/io/pravega/segmentstore/server/ContainerEventProcessor.java

```diff
@@ -0,0 +1,144 @@
+/**
+ * Copyright Pravega Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.pravega.segmentstore.server;
+
+import io.pravega.common.util.BufferView;
+import lombok.Data;
+import lombok.NonNull;
+
+import java.time.Duration;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+
+/**
+ * The {@link ContainerEventProcessor} is a sub-component in a Segment Container that aims at providing a durable,
+ * FIFO-like queue abstraction over an internal, system-critical Segment. The {@link ContainerEventProcessor} can manage
+ * one or more {@link EventProcessor}s, which are the ones that append events to the queue and handle events read. The
+ * also {@link ContainerEventProcessor} reports metrics for all the registered {@link EventProcessor}s.
+ */
+public interface ContainerEventProcessor extends AutoCloseable {
+
+    /**
+     * Instantiates a new {@link EventProcessor}. If the internal Segment exists, the {@link EventProcessor} will re-use
+     * it. If not, a new internal Segment will be created. Multiple calls to this method for the same name should result
+     * in returning the same {@link EventProcessor} object.
+     *
+     * @param name     Name of the {@link EventProcessor} object.
+     * @param handler  Function that will be invoked when one or more events have been read from the internal Segment.
+     * @param config   {@link EventProcessorConfig} for this {@link EventProcessor}.
+     * @return A {@link CompletableFuture} that, when completed, returns a new {@link EventProcessor} object associated
+     * to its own internal Segment.
+     */
+    CompletableFuture<EventProcessor> forConsumer(@NonNull String name,
```

Comment:
It will return the same `EventProcessor` object to all the callers, but I have added a test called `testConcurrentForConsumerCall()` for further testing.

Reply:
But what about the handler ?
Each call potentially provides different handler - how do we handle that?

Reply:
The method is synchronized, so the first thread that completes the execution of this call for the same `name` will succeed and the rest will get the `EventProcessor` with the handler related to the first successful call. Note that having many threads competing to call `forConsumer()` for the _same name and different handlers_ looks wrong in the first place.

Reply:
I am suggesting that we just throw somet Exception for second call that prevents double calling.

Reply:
This is the `forConsumer()` documentation in the API, which was there since the beginning:
```
* Instantiates a new {@link EventProcessor}. If the internal Segment exists, the {@link EventProcessor} will re-use
* it. If not, a new internal Segment will be created. Multiple calls to this method for the same name should result
* in returning the same {@link EventProcessor} object.
```
This function is idempotent, as many other functions in Pravega. Is there any strong reason to change the behavior of it?